xds/server.go

```diff
@@ -0,0 +1,350 @@
+)
+
+var (
+	// These new functions will be overridden in unit tests.
+	newXDSClient = func(opts xdsclient.Options) (xdsClientInterface, error) {
+		return xdsclient.New(opts)
+	}
+	newXDSConfig  = bootstrap.NewConfig
+	newGRPCServer = func(opts ...grpc.ServerOption) grpcServerInterface {
+		return grpc.NewServer(opts...)
+	}
+
+	logger = grpclog.Component("xds")
+)
+
+func prefixLogger(p *GRPCServer) *internalgrpclog.PrefixLogger {
+	return internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(serverPrefix, p))
+}
+
+// xdsClientInterface contains methods from xdsClient.Client which are used by
+// the server. This is useful for overriding in unit tests.
+type xdsClientInterface interface {
+	WatchListener(string, func(xdsclient.ListenerUpdate, error)) func()
+	Close()
+}
+
+// grpcServerInterface contains methods from grpc.Server which are used by the
+// GRPCServer type here. This is useful for overriding in unit tests.
+type grpcServerInterface interface {
+	RegisterService(*grpc.ServiceDesc, interface{})
+	Serve(net.Listener) error
+	Stop()
+	GracefulStop()
+}
+
+// ServeOptions contains parameters to configure the Serve() method.
+//
+// Experimental
+//
+// Notice: This type is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type ServeOptions struct {
+	// Network identifies the local network to listen on. The network must be
+	// "tcp", "tcp4", "tcp6".
+	Network string
+	// Address contains the local address to listen on. This should be of the
+	// form "host:port", where the host must be a literal IP address, and port
+	// must be a literal port number. If the host is a literal IPv6 address it
+	// must be enclosed in square brackets, as in "[2001:db8::1]:80. The host
+	// portion can be left unspecified.
+	Address string
+}
+
+func (so *ServeOptions) validate() error {
+	if net := so.Network; net != "tcp" && net != "tcp4" && net != "tcp6" {
+		return fmt.Errorf("xds: unsupported network type %q for server listener", net)
+	}
+	if _, _, err := net.SplitHostPort(so.Address); err != nil {
+		return fmt.Errorf("xds: unsupported address %q for server listener", so.Address)
+	}
+	return nil
+}
+
+// GRPCServer wraps a gRPC server and provides server-side xDS functionality, by
+// communication with a management server using xDS APIs.
+//
+// Experimental
+//
+// Notice: This type is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type GRPCServer struct {
+	gs     grpcServerInterface
+	quit   *grpcsync.Event
+	logger *internalgrpclog.PrefixLogger
+
+	// clientMu is used only in initXDSClient(), which is called at the
+	// beginning of Serve(), where we have to decide if we have to create a
+	// client or use an existing one.
+	clientMu sync.Mutex
+	xdsC     xdsClientInterface
+}
+
+// NewGRPCServer creates a gRPC server using the passed in ServerOptions. It
+// adds additional stream and unary interceptors to perform any xDS specific
+// functionality, and this is transparent to the users.
```

Comment:
`s/a gRPC server/an xds-enabled gRPC server/`?

Maybe remove the bits about the interceptors as those are implementation details?